src/util.ts

```diff
@@ -769,19 +770,81 @@ export function validateMinimumIncrease(proposed: string, min: string) {
 }
 
 /**
- * Extracts content identifier from ipfs url.
+ * Removes ipfs protocol prefix from ipfs-url.
  *
- * @param url - Ipfs url.
- * @returns Ipfs content identifier as string.
+ * @param ipfsUrl - ipfs url
+ * @returns Ipfs content identifier and (possibly) path in a string
+ * @throws will throw if the url passed is not ipfs.
  */
-export function getIpfsUrlContentIdentifier(url: string): string {
-  if (url.startsWith('ipfs://ipfs/')) {
-    return url.replace('ipfs://ipfs/', '');
-  }
+export function removeIpfsProtocolPrefix(ipfsUrl: string) {
+  if (ipfsUrl.startsWith('ipfs://ipfs/')) {
+    return ipfsUrl.replace('ipfs://ipfs/', '');
+  } else if (ipfsUrl.startsWith('ipfs://')) {
+    return ipfsUrl.replace('ipfs://', '');
+  }
+  // this method should not be used with non-ipfs urls (i.e. startsWith('ipfs://') === true)
+  throw new Error('this method should not be used with non ipfs urls');
+}
 
-  if (url.startsWith('ipfs://')) {
-    return url.replace('ipfs://', '');
+/**
+ * Extracts content identifier and path from ipfs url.
+ *
+ * @param ipfsUrl - ipfs url
+ * @returns Ipfs content identifier as string and path as string.
+ * @throws Will throw if the url passed is not ipfs.
+ */
+export function getIpfsCIDv1AndPath(
+  ipfsUrl: string,
+): { cid: string; path?: string } {
+  const url = removeIpfsProtocolPrefix(ipfsUrl);
+
+  // check if there is a path
+  // (CID is everything preceding first forward slash, path is everything after)
+  const index = url.indexOf('/');
+  const cid = index !== -1 ? url.substring(0, index) : url;
+  const path = index !== -1 ? url.substring(index) : '';
+
+  // we want to ensure that the CID is v1 (https://docs.ipfs.io/concepts/content-addressing/#identifier-formats)
+  return {
+    cid: CID.parse(cid).toV1().toString(),
+    path,
+  };
+}
+
+/**
+ * Adds URL protocol prefix to url string if missing.
+ *
+ * @param urlString - Ipfs url.
+ * @returns string.
+ */
+export function addUrlProtocolPrefix(urlString: string): string {
+  if (!urlString.match(/(^http:\/\/)|(^https:\/\/)/u)) {
+    return `https://${urlString}`;
   }
+  return urlString;
+}
 
-  return url;
+/**
+ * Formats url correctly for use retrieving assets hosted on IPFS.
+ *
+ * @param ipfsGateway - the user preferred ipfsGateway.
+ * @param ipfsUrl - the ipfs url pointed at the asset.
+ * @param subdomainSupported - boolean indicating whether the url should be formatted with subdomains or not
+ * @returns string.
+ */
+export function getFormattedIpfsUrl(
+  ipfsGateway: string,
+  ipfsUrl: string,
+  subdomainSupported: boolean,
+): string {
+  if (subdomainSupported) {
+    const gatewayHost = new URL(addUrlProtocolPrefix(ipfsGateway)).host;
+    const { cid, path } = getIpfsCIDv1AndPath(ipfsUrl);
+    return `https://${cid}.ipfs.${gatewayHost}${path}`;
```

Comment:
It would be better to extract the protocol from the IPFS gateway as well, if it's specified. Users might want to use alternative protocols.

```suggestion
    const { host, protocol } = new URL(addUrlProtocolPrefix(ipfsGateway));
    const { cid, path } = getIpfsCIDv1AndPath(ipfsUrl);
    return `${protocol}//${cid}.ipfs.${host}${path}`;
```

Reply:
done here: https://github.com/MetaMask/controllers/pull/655/commits/d4c8ade79f415cfac20002ae2cc47780f0b44e7a